### PR TITLE
fixes #181: Giving exception if Labels contains space

### LINKED ISSUE #181: Giving exception if Labels contains space
I am getting below error for any syntax error in any query. When looked into code issue seems with the way it is written without quotes in getDatabaseProperties(BoltNeo4jDatabaseMetaData.java:102) while spring jdbc tries to translate the error codes. Connection.getMetaData method always fails if any of the Label in db contains space.

2019-01-22 18:07:55.385 ERROR 11328 --- [nio-8043-exec-1] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is org.neo4j.driver.v1.exceptions.ClientException: Invalid input 'U': expected whitespace, comment, NodeLabel, MapLiteral, a parameter, ')' or a relationship pattern (line 1, column 19 (offset: 18))
"MATCH (n:Business Unit) WITH n LIMIT 1000 UNWIND keys(n) as key RETURN collect(distinct key) as keys"
                   ^] with root cause

org.neo4j.driver.v1.exceptions.ClientException: Invalid input 'U': expected whitespace, comment, NodeLabel, MapLiteral, a parameter, ')' or a relationship pattern (line 1, column 19 (offset: 18))
"MATCH (n:Business Unit) WITH n LIMIT 1000 UNWIND keys(n) as key RETURN collect(distinct key) as keys"
                   ^
	at org.neo4j.driver.internal.net.SocketResponseHandler.handleFailureMessage(SocketResponseHandler.java:76) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.driver.internal.messaging.PackStreamMessageFormatV1$Reader.unpackFailureMessage(PackStreamMessageFormatV1.java:470) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.driver.internal.messaging.PackStreamMessageFormatV1$Reader.read(PackStreamMessageFormatV1.java:431) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.driver.internal.net.SocketClient.receiveOne(SocketClient.java:196) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.driver.internal.net.SocketConnection.receiveOne(SocketConnection.java:217) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.driver.internal.net.ConcurrencyGuardingConnection.receiveOne(ConcurrencyGuardingConnection.java:165) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.driver.internal.net.pooling.PooledSocketConnection.receiveOne(PooledSocketConnection.java:183) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.driver.internal.InternalStatementResult.receiveOne(InternalStatementResult.java:335) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.driver.internal.InternalStatementResult.tryFetchNext(InternalStatementResult.java:325) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.driver.internal.InternalStatementResult.hasNext(InternalStatementResult.java:193) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.jdbc.bolt.BoltNeo4jDatabaseMetaData.cycleResultSetToSetDatabaseProperties(BoltNeo4jDatabaseMetaData.java:109) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.jdbc.bolt.BoltNeo4jDatabaseMetaData.getDatabaseProperties(BoltNeo4jDatabaseMetaData.java:102) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.jdbc.bolt.BoltNeo4jDatabaseMetaData.<init>(BoltNeo4jDatabaseMetaData.java:62) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.jdbc.bolt.impl.BoltNeo4jConnectionImpl.getMetaData(BoltNeo4jConnectionImpl.java:95) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at org.neo4j.jdbc.bolt.impl.BoltNeo4jConnectionImpl.getMetaData(BoltNeo4jConnectionImpl.java:46) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_131]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_131]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_131]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_131]
	at org.neo4j.jdbc.utils.Neo4jInvocationHandler.invoke(Neo4jInvocationHandler.java:54) ~[neo4j-jdbc-driver-3.3.1.jar:na]
	at com.sun.proxy.$Proxy136.getMetaData(Unknown Source) ~[na:na]
	at com.zaxxer.hikari.pool.ProxyConnection.getMetaData(ProxyConnection.java:361) ~[HikariCP-2.7.9.jar:na]
	at com.zaxxer.hikari.pool.HikariProxyConnection.getMetaData(HikariProxyConnection.java) ~[HikariCP-2.7.9.jar:na]
	at org.springframework.jdbc.support.JdbcUtils.extractDatabaseMetaData(JdbcUtils.java:320) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.jdbc.support.JdbcUtils.extractDatabaseMetaData(JdbcUtils.java:356) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.jdbc.support.SQLErrorCodesFactory.getErrorCodes(SQLErrorCodesFactory.java:214) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.jdbc.support.SQLErrorCodeSQLExceptionTranslator.setDataSource(SQLErrorCodeSQLExceptionTranslator.java:136) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.jdbc.support.SQLErrorCodeSQLExceptionTranslator.<init>(SQLErrorCodeSQLExceptionTranslator.java:99) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.jdbc.support.JdbcAccessor.getExceptionTranslator(JdbcAccessor.java:122) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1402) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:620) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:657) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:682) ~[spring-jdbc-5.0.7.RELEASE.jar:5.0.7.RELEASE]

### PR DESCRIPTION
**Pull request by conker84**
Changes:
- Fixed metadata resolution in case of backtick
- Fixed node property resolution in case of a property contains a backtick
- Fixed relationship property resolution in case of a property contains a backtick

### CODE REVIEW

File: neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltNeo4jResultSet.java [hunk 1]
```diff
@@ -354,7 +367,7 @@ private Value fetchPropertyValue(String key, String property) throws SQLExceptio
 				value = new StringValue(this.current.get(key).asRelationship().type());
 			} else {
 				//Property requested
-				value = this.current.get(key).get(property);
+				value = this.current.get(key).get(property.replaceAll("`", ""));
```
Comment (jexp) [inline neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltNeo4jResultSet.java]: move into a santiize function and use across labels, rel-types and property-keys

**Comment by conker84** [comment]
@jexp updated!

File: neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltNeo4jResultSet.java [hunk 2]
```diff
@@ -60,6 +63,11 @@
 
 	private static final List<String> ACCEPTED_TYPES_FOR_FLATTENING = Arrays.asList("NODE", "RELATIONSHIP");
 
+	private static final Pattern SPECIAL_CHARS = Pattern.compile("[^a-z0-9]", Pattern.CASE_INSENSITIVE);
```
Comment (jexp) [inline neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltNeo4jResultSet.java]: misses underscore

also perhaps a positive pattern would be better? i.e. `^[a-z_][a-z0-9_]+$` + case insentistive ?

File: neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltNeo4jResultSet.java [hunk 3]
```diff
@@ -143,13 +151,19 @@ private void flattenNode(Node node, String nodeKey) {
...
+			String newKey = parsePropertyKey(nodeKey, key);
+			if (keys.indexOf(newKey) == -1) {
+				keys.add(newKey);
 				classes.add(node.get(key).type());
 			}
 		}
 	}
 
+	private String parsePropertyKey(String nodeKey, String key) {
+		String keyTemplate = SPECIAL_CHARS.matcher(key).find() ? FLATTEN_KEY_TEMPLATE_BACKTICK : FLATTEN_KEY_TEMPLATE;
```
Comment (jexp) [inline neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltNeo4jResultSet.java]: shouldn't this be part of a `quote` method?

I think that would make it easier, like I do it in other places:

```
def quote (name:String) = if (name.isSpecial()) '`'+sanitize(name)+'`' else name
```

and use that in all places instead all the string formats?

Comment (jexp) [reply neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltNeo4jResultSet.java]: same for labels and rel-types

Comment (jexp) [reply neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltNeo4jResultSet.java]: isSpecial could also use the 

Character.isJavaIdentifierStart(c) 
Character.isJavaIdentifier(c)

instaed of regexp

**Review by jexp — CHANGES_REQUESTED** [review]

**Comment by conker84** [comment]
@jexp updated!

**Review by jexp — APPROVED** [review]

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/utils/SanitizeUtils.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * This method return a quoted String with backtick (<b>`</b>) if the input is not a valid java identifier, otherwise
+     * returns the string itself
+     * <br>
+     * i.e. input: property name => output => `property name`
+     *
+     * @param raw The string to be quoted.
+     * @return The quoted string.
+     */
+    public static String quote(String value) {
```
Comment (jexp) [inline neo4j-jdbc/src/main/java/org/neo4j/jdbc/utils/SanitizeUtils.java:55]: shouldn't we use this in more places?

Comment (jexp) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/utils/SanitizeUtils.java:55]: I.e. whenever we get a property key / label / rel-type that we add to a cypher statement?
 Also it should use `sanitize()` for the value before wrapping it in backticks.

**Pull request merged**